modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/writer/interceptor/ObjectNestedFieldsWriterInterceptor.java

```diff
@@ -0,0 +1,248 @@
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.UriInfo;
+import javax.ws.rs.ext.WriterInterceptor;
+import javax.ws.rs.ext.WriterInterceptorContext;
+
+import org.apache.cxf.jaxrs.ext.ContextProvider;
+import org.apache.cxf.jaxrs.impl.UriInfoImpl;
+import org.apache.cxf.jaxrs.provider.ProviderFactory;
+import org.apache.cxf.message.Message;
+
+/**
+ * @author Javier de Arcos
+ */
+public class ObjectNestedFieldsWriterInterceptor implements WriterInterceptor {
+
+	public ObjectNestedFieldsWriterInterceptor(
+		List<ObjectRelationship> objectRelationships) {
+
+		_objectRelationships = objectRelationships;
+	}
+
+	@Override
+	public void aroundWriteTo(WriterInterceptorContext writerInterceptorContext)
+		throws IOException, WebApplicationException {
+
+		Object entity = writerInterceptorContext.getEntity();
+
+		if (!(entity instanceof ObjectEntry)) {
+			writerInterceptorContext.proceed();
+
+			return;
+		}
+
+		NestedFieldsContext nestedFieldsContext =
+			NestedFieldsContextThreadLocal.getNestedFieldsContext();
+
+		if (nestedFieldsContext == null) {
+			writerInterceptorContext.proceed();
+
+			return;
+		}
+
+		try {
+			_setFieldValues(
+				(ObjectEntry)entity, nestedFieldsContext.getFieldNames(),
+				nestedFieldsContext);
+
+			_removeRelationIdProperties((ObjectEntry)entity);
+		}
+		catch (Exception exception) {
+			_log.error(exception.getMessage(), exception);
+
+			throw new WebApplicationException(exception);
+		}
+
+		writerInterceptorContext.proceed();
+	}
+
+	protected ProviderFactory getProviderFactory(Message message) {
+		return ProviderFactory.getInstance(message);
+	}
+
+	private <T> Object _getContext(Class<T> contextClass, Message message) {
+		if (contextClass.equals(UriInfo.class)) {
+			return new UriInfoImpl(message);
+		}
+
+		ContextProvider<?> contextProvider = _getContextProvider(
+			contextClass, message);
+
+		if (contextProvider != null) {
+			return contextProvider.createContext(message);
+		}
+
+		return null;
+	}
+
+	private <T> ContextProvider<T> _getContextProvider(
+		Class<T> contextClass, Message message) {
+
+		ProviderFactory providerFactory = getProviderFactory(message);
+
+		return providerFactory.createContextProvider(contextClass, message);
+	}
+
+	private ObjectEntry _getNestedFieldValue(
+			String fieldName, ObjectEntry objectEntry,
+			NestedFieldsContext nestedFieldsContext)
+		throws Exception {
+
+		ObjectEntryResource objectEntryResource = ObjectEntryResource.builder(
+		).user(
+			(User)_getContext(User.class, nestedFieldsContext.getMessage())
+		).build();
+
+		_setResourceContexts(
+			nestedFieldsContext.getMessage(), objectEntryResource);
+
+		Long objectEntryId = _getRelatedObjectEntryId(fieldName, objectEntry);
+
+		if (objectEntryId == null) {
+			return null;
+		}
+
+		return objectEntryResource.getObjectEntry(objectEntryId);
```

Comment:
Do we need the resource here? How do we get an ObjectDefinition?

Reply:
> You can not generate directly the resource files, as you can not inject the resource directly (you need to inject the factory component). I prepared a PR about it (https://issues.liferay.com/browse/LPS-134176) but did not sent it, let me see if I can restore it.

Removing the Prototype scope made it work, but I guess I didn't consider some other cases... The good news is that this is something we have in our radar to achieve and even better if you had something prepared for this

Reply:
Does not appl with the ObjectEntryDTO approach